oml/losses/arcface.py

```diff
@@ -0,0 +1,130 @@
+                not category-wise
+            label_smoothing: Label smoothing effect strength
+            m: Margin parameter for ArcFace loss. Usually you should use 0.3-0.5 values for it
+            s: Scaling parameter for ArcFace loss. Usually you should use 30-64 values for it
+        """
+        super(ArcFaceLoss, self).__init__()
+
+        assert (
+            label_smoothing is None or 0 < label_smoothing < 1
+        ), f"Choose another label_smoothing parametrization, got {label_smoothing}"
+
+        self.criterion = nn.CrossEntropyLoss()
+        self.num_classes = num_classes
+        if label2category is not None:
+            mapper = {l: i for i, l in enumerate(sorted(list(set(label2category.values()))))}
+            label2category = {k: mapper[v] for k, v in label2category.items()}
+            self.label2category = torch.arange(num_classes).apply_(label2category.get)
+        else:
+            self.label2category = None
+        self.label_smoothing = label_smoothing
+        self.weight = nn.Parameter(torch.FloatTensor(num_classes, in_features))
+        nn.init.xavier_uniform_(self.weight)
+        self.rescale = s
+        self.m = m
+        self.cos_m = np.cos(m)
+        self.sin_m = np.sin(m)
+        self.th = -self.cos_m
+        self.mm = self.sin_m * m
+        self.last_logs: Dict[str, float] = {}
+
+    def fc(self, x: torch.Tensor) -> torch.Tensor:
+        return F.linear(F.normalize(x, p=2), F.normalize(self.weight, p=2))
+
+    def smooth_labels(self, y: torch.Tensor) -> torch.Tensor:
+        if self.label2category is not None:
+            self.label2category = self.label2category.to(self.weight.device)
+        return label_smoothing(y, self.num_classes, self.label_smoothing, self.label2category)
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        assert torch.all(y < self.num_classes), "You should provide labels between 0 and num_classes - 1."
+
+        cos = self.fc(x)
+
+        self._log_accuracy_on_batch(cos, y)
+
+        sin = torch.sqrt(1.0 - torch.pow(cos, 2))
+
+        cos_w_margin = cos * self.cos_m - sin * self.sin_m
+        cos_w_margin = torch.where(cos > self.th, cos_w_margin, cos - self.mm)
+
+        ohe = F.one_hot(y, self.num_classes)
+        logit = torch.where(ohe.bool(), cos_w_margin, cos) * self.rescale
+
+        if self.label_smoothing:
+            y = self.smooth_labels(y)
+
+        return self.criterion(logit, y)
+
+    @torch.no_grad()
+    def _log_accuracy_on_batch(self, logits: torch.Tensor, y: torch.Tensor) -> torch.Tensor:
+        self.last_logs["accuracy"] = torch.mean((y == torch.argmax(logits, 1)).to(torch.float32))
+
+
+class ArcFaceLossWithMLP(nn.Module):
+    """
+    Almost the same as ``ArcFaceLoss``, but also has MLP projector before the loss.
+    We need this because all layers will be removed from loss at test time.
```

Comment:
`We need this because all layers will be removed from loss at test time.`
->
```
You may want to use `ArcFaceLossWithMLP` to boost the expressive power of ArcFace loss during the training (for example, in a multi-head setup it may be a good idea to have task-specific projectors in each of the losses).

Note, the criterion does not exist during the validation time.
Thus, if you want to keep your MLP layers, you should create them as a part of the model you train.
```

Reply:
rephrase ^